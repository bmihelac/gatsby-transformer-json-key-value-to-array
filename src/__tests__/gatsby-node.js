const { onCreateNode } = require(`../gatsby-node`)

// Make some fake functions its expecting.
const loadNodeContent = node => Promise.resolve(node.content)

const bootstrapTest = async (node, pluginOptions = {}) => {
  const createNode = jest.fn()
  const createParentChildLink = jest.fn()
  const actions = { createNode, createParentChildLink }
  const createNodeId = jest.fn()
  createNodeId.mockReturnValue(`uuid-from-gatsby`)
  const createContentDigest = jest.fn().mockReturnValue(`contentDigest`)

  return await onCreateNode(
    {
      node,
      loadNodeContent,
      actions,
      createNodeId,
      createContentDigest,
    },
    pluginOptions
  ).then(() => {
    return {
      createNode,
      createParentChildLink,
    }
  })
}

describe(`Process JSON nodes correctly`, () => {
  const baseNode = {
    id: `whatever`,
    parent: `SOURCE`,
    children: [],
    internal: {
      contentDigest: `whatever`,
      mediaType: `application/json`,
    },
  }

  const baseFileNode = {
    ...baseNode,
    name: `nodeName`,
    dir: `/tmp/foo/`,
    internal: {
      ...baseNode.internal,
      type: `File`,
    },
  }

  it(`correctly creates nodes from JSON`, async () => {
    const data = {
      title: 'My title',
      description: 'Lorem ipsum',
      nested: {foo: 'bar'},
    }
    const node = {
      ...baseFileNode,
      content: JSON.stringify(data),
    }

    return bootstrapTest(node).then(({ createNode, createParentChildLink }) => {
      expect(createNode.mock.calls).toMatchSnapshot()
      expect(createParentChildLink.mock.calls).toMatchSnapshot()
      expect(createNode).toHaveBeenCalledTimes(3)
      expect(createParentChildLink).toHaveBeenCalledTimes(3)
    })
  })
})
