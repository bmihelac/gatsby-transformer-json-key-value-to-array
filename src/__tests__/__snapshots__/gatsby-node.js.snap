// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Process JSON nodes correctly correctly creates nodes from JSON 1`] = `
Array [
  Array [
    Object {
      "children": Array [],
      "file___NODE": "whatever",
      "id": "uuid-from-gatsby",
      "internal": Object {
        "contentDigest": "contentDigest",
        "type": "KeyValue",
      },
      "isJson": false,
      "key": "title",
      "parent": "whatever",
      "value": "My title",
    },
  ],
  Array [
    Object {
      "children": Array [],
      "file___NODE": "whatever",
      "id": "uuid-from-gatsby",
      "internal": Object {
        "contentDigest": "contentDigest",
        "type": "KeyValue",
      },
      "isJson": false,
      "key": "description",
      "parent": "whatever",
      "value": "Lorem ipsum",
    },
  ],
  Array [
    Object {
      "children": Array [],
      "file___NODE": "whatever",
      "id": "uuid-from-gatsby",
      "internal": Object {
        "contentDigest": "contentDigest",
        "type": "KeyValue",
      },
      "isJson": true,
      "key": "nested",
      "parent": "whatever",
      "value": "{\\"foo\\":\\"bar\\"}",
    },
  ],
]
`;

exports[`Process JSON nodes correctly correctly creates nodes from JSON 2`] = `
Array [
  Array [
    Object {
      "child": Object {
        "children": Array [],
        "file___NODE": "whatever",
        "id": "uuid-from-gatsby",
        "internal": Object {
          "contentDigest": "contentDigest",
          "type": "KeyValue",
        },
        "isJson": false,
        "key": "title",
        "parent": "whatever",
        "value": "My title",
      },
      "parent": Object {
        "children": Array [],
        "content": "{\\"title\\":\\"My title\\",\\"description\\":\\"Lorem ipsum\\",\\"nested\\":{\\"foo\\":\\"bar\\"}}",
        "dir": "/tmp/foo/",
        "id": "whatever",
        "internal": Object {
          "contentDigest": "whatever",
          "mediaType": "application/json",
          "type": "File",
        },
        "name": "nodeName",
        "parent": "SOURCE",
      },
    },
  ],
  Array [
    Object {
      "child": Object {
        "children": Array [],
        "file___NODE": "whatever",
        "id": "uuid-from-gatsby",
        "internal": Object {
          "contentDigest": "contentDigest",
          "type": "KeyValue",
        },
        "isJson": false,
        "key": "description",
        "parent": "whatever",
        "value": "Lorem ipsum",
      },
      "parent": Object {
        "children": Array [],
        "content": "{\\"title\\":\\"My title\\",\\"description\\":\\"Lorem ipsum\\",\\"nested\\":{\\"foo\\":\\"bar\\"}}",
        "dir": "/tmp/foo/",
        "id": "whatever",
        "internal": Object {
          "contentDigest": "whatever",
          "mediaType": "application/json",
          "type": "File",
        },
        "name": "nodeName",
        "parent": "SOURCE",
      },
    },
  ],
  Array [
    Object {
      "child": Object {
        "children": Array [],
        "file___NODE": "whatever",
        "id": "uuid-from-gatsby",
        "internal": Object {
          "contentDigest": "contentDigest",
          "type": "KeyValue",
        },
        "isJson": true,
        "key": "nested",
        "parent": "whatever",
        "value": "{\\"foo\\":\\"bar\\"}",
      },
      "parent": Object {
        "children": Array [],
        "content": "{\\"title\\":\\"My title\\",\\"description\\":\\"Lorem ipsum\\",\\"nested\\":{\\"foo\\":\\"bar\\"}}",
        "dir": "/tmp/foo/",
        "id": "whatever",
        "internal": Object {
          "contentDigest": "whatever",
          "mediaType": "application/json",
          "type": "File",
        },
        "name": "nodeName",
        "parent": "SOURCE",
      },
    },
  ],
]
`;
